src/config.rs

```diff
@@ -0,0 +1,376 @@
+    ///
+    /// # Arguments
+    /// * `config_path` - Path to the configuration file
+    fn new_from_config_file(config_path: &Path) -> miette::Result<Self> {
+        let config = Self::parse_config(config_path)?;
+
+        // Load edition from root of the config file
+        let edition = config
+            .get("edition")
+            .and_then(|edition| edition.as_str())
+            .ok_or_else(|| miette!("missing or invalid 'edition' field in config file"))?
+            .into();
+
+        match edition {
+            Edition::Canary => (),
+            _ => bail!("unsupported config file edition, supported editions: {CANARY_EDITION}"),
+        }
+
+        // Load registries from [registries] section
+        let registries = Self::get_registries(&config, config_path)?;
+
+        // Locate default registry from [registry.default]
+        let default_registry = Self::get_default_registry(&config, &registries)?;
+
+        // Parse command-specific default arguments from [commands.*] sections
+        let command_defaults = Self::get_command_defaults(config, config_path)?;
+
+        Ok(Self {
+            edition,
+            config_path: Some(config_path.to_owned()),
+            default_registry,
+            registries,
+            command_defaults,
+        })
+    }
+
+    fn parse_config(config_path: &Path) -> Result<toml::Value, miette::Error> {
+        let config_str = std::fs::read_to_string(config_path)
+            .into_diagnostic()
+            .wrap_err(miette!(
+                "failed to read config file: {}",
+                config_path.display()
+            ))?;
+        let config: toml::Value =
+            toml::from_str(&config_str)
+                .into_diagnostic()
+                .wrap_err(miette!(
+                    "failed to parse config file: {}",
+                    config_path.display()
+                ))?;
+        Ok(config)
+    }
+
+    fn get_default_registry(
+        config: &toml::Value,
+        registries: &HashMap<String, RegistryUri>,
+    ) -> miette::Result<Option<String>> {
+        let default_registry = config
+            .get("registry")
+            .and_then(|registry| registry.get("default"))
+            .and_then(|default| default.as_str())
+            .map(|default| default.to_string());
+        if let Some(ref default_registry) = default_registry {
+            ensure!(
+                registries.contains_key(default_registry),
+                "default registry '{}' not found in list of registries",
+                default_registry
+            );
+        }
+        Ok(default_registry)
+    }
+
+    /// Load registries from the [registries] section of the config file
+    ///
+    /// # Arguments
+    /// * `config` - The parsed TOML configuration
+    /// * `config_path` - Path to the configuration file (for error messages)
+    ///
+    /// # Returns
+    /// A map of registry names to URIs
+    fn get_registries(
```

Comment:
(Related to the serde comment)

You can save a ton of code and error surface + following idiomatic language patterns by doing a quick serde impl here. For me this would be the last blocker to merging this though!

Reply:
e48abf21eab4565706ab4f078a2a38f14af05f3f